pkgs/cli_config/lib/src/config.dart

```diff
@@ -0,0 +1,387 @@
+  /// If provided, [environment] must be a map containing environment varibales.
+  ///
+  /// At most one of [fileContents] or [fileParsed] must be provided.
+  /// If provided, [fileContents] must be valid JSON or YAML.
+  /// If provided, [fileParsed] must be valid parsed YSON or YAML (maps, lists,
+  /// strings, integers, and booleans).
+  ///
+  /// If provided [fileSourceUri], is used to provide better error messages on
+  /// parsing the configuration file.
+  factory Config({
+    List<String> cliDefines = const [],
+    Map<String, String> environment = const {},
+    Map<String, dynamic>? fileParsed,
+    String? fileContents,
+    Uri? fileSourceUri,
+  }) {
+    // Parse config file.
+    if (_countNonNulls([fileContents, fileParsed]) > 1) {
+      throw ArgumentError(
+          'Provide at most one of `fileParsed` and `fileContents`.');
+    }
+    final Map<String, dynamic> fileConfig;
+    if (fileParsed != null) {
+      fileConfig = FileParser().parseMap(fileParsed);
+    } else if (fileContents != null) {
+      fileConfig = FileParser().parse(
+        fileContents,
+        sourceUrl: fileSourceUri,
+      );
+    } else {
+      fileConfig = {};
+    }
+
+    // Parse CLI argument defines.
+    final cliConfig = DefinesParser().parse(cliDefines);
+
+    // Parse environment.
+    final environmentConfig = EnvironmentParser().parse(environment);
+
+    return Config._(
+      CliProvider(cliConfig),
+      EnvironmentProvider(environmentConfig),
+      FileProvider(fileConfig, fileSourceUri),
+    );
+  }
+
+  /// Constructs a config by parsing CLI arguments and loading the config file.
+  ///
+  /// The [args] must be commandline arguments.
+  ///
+  /// If provided, [environment] must be a map containing environment varibales.
+  /// If not provided, [environment] defaults to [Platform.environment].
+  ///
+  /// This async constructor is intended to be used directly in CLI files.
+  static Future<Config> fromArgs({
+    required List<String> args,
+    Map<String, String>? environment,
+  }) async {
+    final results = CliParser().parse(args);
+
+    // Load config file.
+    final configFile = results['config'] as String?;
+    String? fileContents;
+    Uri? fileSourceUri;
+    if (configFile != null) {
+      fileContents = await File(configFile).readAsString();
+      fileSourceUri = Uri.file(configFile);
+    }
+
+    return Config(
+      cliDefines: results['define'] as List<String>,
+      environment: environment ?? Platform.environment,
+      fileContents: fileContents,
+      fileSourceUri: fileSourceUri,
+    );
+  }
+
+  /// Lookup a string value in this config.
+  ///
+  /// First tries CLI argument defines, then environment variables, and
+  /// finally the config file.
+  ///
+  /// Throws if one of the configs does not contain the expected value type.
+  ///
+  /// If [validValues] is provided, throws if an unxpected value is provided.
+  String getString(String key, {Iterable<String>? validValues}) {
```

Comment:
It feels like we might want to be a bit more general and just support a validation callback?

Reply:
> The main thing is right now is by far the easiest time to modify the API, and bolting on different ways of doing validation later is likely going to just make for a cludgy API with lots of ways of doing the same thing.

Agreed. We should deprecate and remove it if we want anther solution before releasing a v1.0.
At the same time, I don't want to already solve all potential issues in v0.1.